### PR TITLE
fix: use list pay instead of list invoices modified

### LINKED ISSUE #28: use listpays instead of listinvoices in the home view
List invoices did not contain the payments that we did or received with the node, but the listpays does.

In addition, payment can decrease or increase the actual capacity of the node and this is something that we can calculate with the node id contained inside the from the field of the `listpays` command

### PR DESCRIPTION
**Pull request by Harshit933**
For now the screen shows all the transaction that are completed or failed following the command :

`lightning-cli listpays --testnet`

Screenshot from 2023-03-09 23-44-35

fix #28

### CODE REVIEW

**Comment by vincenzopalazzo** [comment]
we can add only the payment that was completed with success?

You need to pass the status=complete

**Comment by Harshit933** [comment]
> we can add only the payment that was completed with success?
> 
> You need to pass the status=complete

Yes we can modify that, just let me know if this way is good to go or am I missing something here.

File: lib/api/cln/cln_client.dart [hunk 1]
```diff
@@ -161,4 +163,26 @@ class CLNApi extends AppApi {
...
+
+  @override
+  Future<AppListPays> listPays() {
+    dynamic params;
+    switch (mode) {
+      case ClientMode.grpc:
+        params = CLNListPaysRequest(grpcRequest: ListpaysRequest());
+        break;
+      case ClientMode.unixSocket:
+        params = CLNListPaysRequest(unixRequest: <String, dynamic>{});
```
Comment (vincenzopalazzo) [inline lib/api/cln/cln_client.dart:175]:
```suggestion
        params = CLNListPaysRequest(unixRequest: <String, dynamic>{"status": "completed"});
```

Also in the others kind of request,

In addition I do not know if this is the right syntax, so you need to double check

**Comment by vincenzopalazzo** [comment]
>Yes we can modify that, just let me know if this way is good to go or am I missing something here.

Looks like the good direction

File: lib/api/cln/cln_client.dart (hunk 1 above)
Comment (Harshit933) [reply lib/api/cln/cln_client.dart:175]: Sure, I'll double check it

**Comment by Harshit933** [comment]
Actually we can't specify the status of the payment in the `lightning-cli listinvoices --testnet` command. I was stuck here for quite a while. I have figured out a way in the `home_view.dart`. Checking whether the status of the invoice is paid or not.

**Comment by vincenzopalazzo** [comment]
>Actually we can't specify the status of the payment in the lightning-cli listinvoices --testnet command. I was stuck here for quite a while. I have figured out a way in the home_view.dart. Checking whether the status of the invoice is paid or not.

yeah it is not possible, in this case you should:

- open a issue to ask the support on core lightning
- fix your solution by filtering and remove all the invoice that are not paid!

Usually this kind of work need to be done with the core lightning docs to the side

**Comment by Harshit933** [comment]
> * open a issue to ask the support on core lightning
> * fix your solution by filtering and remove all the invoice that are not paid!

Yes, I have filtered the invoices which are not paid.

File: lib/api/api.dart [hunk 2]
```diff
@@ -4,6 +4,8 @@ import 'package:clnapp/model/app_model/list_invoices.dart';
 import 'package:clnapp/model/app_model/list_transaction.dart';
 import 'package:clnapp/model/app_model/pay_invoice.dart';
 
+import '../model/app_model/list_pays.dart';
```
Comment (vincenzopalazzo) [inline lib/api/api.dart:7]:
```suggestion
import 'package:clnapp/model/model/app_model/list_pays.dart';
```

I do not like the local import, sorry

File: lib/views/home/home_view.dart [hunk 3]
```diff
@@ -114,12 +114,15 @@ class _HomeViewState extends State<HomeView> {
 
   Future<List<dynamic>?> listPayments() async {
     final invoicesList = await widget.provider.get<AppApi>().listInvoices();
```
Comment (vincenzopalazzo) [inline lib/views/home/home_view.dart:116]:
```suggestion
    final invoicesList = await widget.provider.get<AppApi>().listInvoices(status: "paid");
```

File: lib/views/home/home_view.dart [hunk 4]
```diff
@@ -158,6 +164,14 @@ class _HomeViewState extends State<HomeView> {
...
                 itemCount: snapshot.data!.length,
                 itemBuilder: (context, index) {
+                  /// Returning a null widget when the invoice is unpaid
+                  if (snapshot.data![index].identifier == "invoice" &&
+                      snapshot.data![index].status == "unpaid") {
+                    return const SizedBox(
+                      width: 0,
+                      height: 0,
+                    );
+                  }
```
Comment (vincenzopalazzo) [inline lib/views/home/home_view.dart:174]: this is the wrong place where to filter, you should add the code inside the `listInvoices` and pass the status `paid` to the method

We need to show the money that we had, so the unpaid invoice are not the right one, but the paid one are

**Review by vincenzopalazzo — CHANGES_REQUESTED** [review]

File: lib/api/api.dart [hunk 5]
```diff
@@ -28,7 +28,7 @@ abstract class AppApi {
   Future<AppListFunds?> listFunds();
 
   /// Return the list of invoices from lightning node.
-  Future<AppListInvoices> listInvoices();
+  Future<AppListInvoices?> listInvoices({required String status});
```
Comment (vincenzopalazzo) [inline lib/api/api.dart:31]:
```suggestion
  Future<AppListInvoices> listInvoices({required String status});
```

why a nullable value?

File: lib/model/app_model/list_invoices.dart [hunk 6]
```diff
@@ -2,18 +2,24 @@ import 'package:cln_common/cln_common.dart';
...
 
 class AppListInvoices {
-  List<AppInvoice> invoice;
+  List<AppInvoice?> invoice;
 
   AppListInvoices({this.invoice = const []});
 
   factory AppListInvoices.fromJSON(Map<String, dynamic> json,
-      {bool snackCase = false, bool isObject = false}) {
+      {bool snackCase = false, bool isObject = false, required String status}) {
```
Comment (vincenzopalazzo) [inline lib/model/app_model/list_invoices.dart:10]:
```suggestion
      {bool snackCase = false, bool isObject = false, required String status}) {
```

why this required String?

File: lib/model/app_model/list_invoices.dart [hunk 7]
```diff
@@ -2,18 +2,24 @@ import 'package:cln_common/cln_common.dart';
 import 'package:clnapp/model/app_model/app_utils.dart';
 
 class AppListInvoices {
-  List<AppInvoice> invoice;
+  List<AppInvoice?> invoice;
```
[the remark this answers is not in the document]
Comment (Harshit933) [reply lib/model/app_model/list_invoices.dart:5]: In the file `list_invoices.dart`. We are checking the status of the transaction to be paid or not. And if it is paid then only we are adding it into the `appInvoices` list, thus we can have a scenario where `appInvoices` could be null.

File: lib/api/api.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (Harshit933) [reply lib/api/api.dart:31]: Changed it

File: lib/views/home/home_view.dart [hunk 8]
```diff
@@ -113,12 +113,15 @@ class _HomeViewState extends State<HomeView> {
...
-    final invoicesList = await widget.provider.get<AppApi>().listInvoices();
+    final invoicesList =
+        await widget.provider.get<AppApi>().listInvoices(status: "paid");
     final paysList = await widget.provider.get<AppApi>().listPays();
 
     List list = [];
 
-    list.addAll(invoicesList.invoice);
+    if (invoicesList != null) {
+      list.addAll(invoicesList.invoice);
```
[the remark this answers is not in the document]
Comment (Harshit933) [reply lib/views/home/home_view.dart]: Thank you changed it

File: lib/model/app_model/list_invoices.dart (hunk 7 above)
[the remark this answers is not in the document]
Comment (vincenzopalazzo) [reply lib/model/app_model/list_invoices.dart:5]: what is the difference to return null instead an appInvoice with a empty list?

**Review by vincenzopalazzo — CHANGES_REQUESTED** [review]

File: lib/model/app_model/list_invoices.dart [hunk 9]
```diff
@@ -13,14 +13,14 @@ class AppListInvoices {
       var appInvoices = invoices.map((invoice) {
         var temp = AppInvoice.fromJSON(invoice,
             snackCase: snackCase, isObject: isObject);
-        if (temp.status == status) {
+        if (temp.status == "unpaid") {
```
Comment (vincenzopalazzo) [inline lib/model/app_model/list_invoices.dart]:
```suggestion
        if (temp.status == "unpaid") {
```

why? this is wrong `AppListInvoices` need to have a status

**Review by vincenzopalazzo — CHANGES_REQUESTED** [review]

File: lib/model/app_model/list_invoices.dart (hunk 9 above)
Comment (Harshit933) [reply lib/model/app_model/list_invoices.dart]: I accidentally changed that to unpaid. I have fixed it in the later commit.

**Comment by vincenzopalazzo** [comment]
Running this code locally throws the following exception

```
flutter: ┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
flutter: │ #0   LogManager.debug (package:cln_common/src/logger/logger_manager.dart:16:13)
flutter: │ #1   RPCClient.call (package:clightning_rpc/src/clightning_rpc.dart:36:10)
flutter: ├┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄
flutter: │ 🐛 Response from RPC is : {pays: [{destination: 03933884aaf1d6b108397e5efe5c86bcf2d8ca8d2f700eda99db9214fc2712b134, payment_hash: 08b814811c94728d59ee04a3d6f3ee23c92a81e48fa5a8e23972f050989058b5, status: complete, created_at: 1616175716, completed_at: 1616175716, preimage: a2effbf3361ad12dd14cb3091d56e9d6566ca5f3d999291fc8b2c1a1a4f30a2e, amount_msat: 200000, amount_sent_msat: 201004}, {bolt11: lntb1n1psxflaqpp5vcvc9uwtnw9flw6gadp2mv0ry4y8jj5cyul6lg66f84y3xvqm32sdq8w3jhxaqxqyjw5qcqp2sp50n6my4zyenrf2ddc0trhlc7g8kyuukp8mhgmp6eecwww5zpjuw4srzjqv9ks6sk823thgpuawut4dmh37kz29fkfxq5rhc2gdkk3q6j6sn0v8v7r5qqqfqqqqqqqqq2qqqq86qqqc9qy9qsq8999a06st5f9llry542rdn9zathwss832kvsd5kqwg0xxh2cawwzhuhw36ypw6zmkxzplpljmg0pwayzmfxtetgcp9q783e23xerkygqtghmwm, destination: 035f3c2d601d3dda4f717409cef7c76e1bdce353d79fbe630693bff32beaf217bc, payment_hash: 661982f1cb9b8a9fbb48eb42adb1e32548794a98273fafa35a49ea489980dc55, status: complete, created_at: 1617231911, completed_at: 1617231911, preimage: 9e2fad8afbeb8857459f59154827b07cdf9d39e5c24fb0ac69a7dae7a6d58568, amount_msat: 100, amount_sent_msat: 110}, {destination: 03b39d1ddf13ce486de74e9e44e0538f960401a9ec75534ba9cfe4100d65426880, payment_hash: 19e149210c2e4b7beb68b15f648b1973b7d9b83f25d434831650966b27cba606, status: complete, created_at: 1651492299, completed_at: 1651492299, preimage: a7f511c1180111ccb05d4eb9760187d61f4bd7cd44c2832aaa06d5592078c7f5, amount_msat: 450000000, amount_sent_msat: 450002901}, {bolt11: lntb1p3ttrc2sp5j5arcf6vfmju9m2y954qkff3vduvye88h4yz5yt49gllemm8qddqpp56g8cdkf9ck6l7q4u74s5lee3zqkwpsz2vjea77qn37md6sf35z8qdqav4ehg6twvusxcm3qwpshjmt9de68xxqyjw5qcqp2rzjqgcjvfla7plmm4l9mkcnvcgmmh5mqrfxsgw3fk2gjyu4g5hk0teysgksjcqqq2sqqyqqqqlgqqqqqqgq9q9qyysgqfw7hyzh66echsxpahqhu2r7zjrurcvq7080r0c9z0zwqxtt6fwkyvtte722v75fqnh2f67vteflw8xxgckrgf93jw59semn6qvku24sptvqkmg, destination: 02ec99e98f3f3a5e300bebb12a49d5b72d77a55796808a8f5b2e6b42f3db51c7cf, payment_hash: d20f86d925c5b5ff02bcf5614fe731102ce0c04a64b3df78138fb6dd4131a08e, status: complete, created_at: 1656068042, completed_at: 1656068042, preimage: aaf19dd5bb5f49cd44e472cad6c185af715011dd9c8c93fc07b87fce38be6100, amount_msat: 100, amount_sent_msat: 100}, {bolt11: lntb30u1p30ryqupp5s48g366yl3aul6yes3fqh05k2tpddrthepp0244n3ne7clqxwkhsdq5xvsyymr0ddskxcmfdehsxqrrsscqp79qy9qsqsp5y2c2kn9u7gdxyafxmw8exu05krcqx5yva54jp7kayz7k6jat3nnqqkf34gwennur2t7de7g8nndq9czlff6q38k27kgjufwu7jthpa4qj426edh3peqhfth36etuk7tcx3hax5uche2ues99vparl8rzpgspu9wu8v, destination: 0348cc1a9479697cd52db445ea74149ad40bb01bb2045a3e8acba21b70f94ab7cf, payment_hash: 854e88eb44fc7bcfe89984520bbe9652c2d68d77c842f556b38cf3ec7c0675af, status: complete, created_at: 1660000901, completed_at: 1660000901, preimage: 38ee9c38eee6f8cc636383bab3d4d0ad87c084dc0e4fd10d71f886059cc58c01, amount_msat: 3000000, amount_sent_msat: 3002503}]}
flutter: └───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
flutter: ┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
flutter: │ #0   LogManager.error (package:cln_common/src/logger/logger_manager.dart:20:13)
flutter: │ #1   _HomeViewState._checkIfThereAreError (package:clnapp/views/home/home_view.dart:218:30)
flutter: ├┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄
flutter: │ ⛔ type 'Null' is not a subtype of type 'String'
flutter: └───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
flutter: ┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
flutter: │ #0   LogManager.error (package:cln_common/src/logger/logger_manager.dart:20:13)
flutter: │ #1   _HomeViewState._checkIfThereAreError (package:clnapp/views/home/home_view.dart:219:30)
flutter: ├┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄
flutter: │ ⛔ #0      new AppPays.fromJSON (package:clnapp/model/app_model/list_pays.dart:58:17)
flutter: │ ⛔ #1      new AppListPays.fromJSON.<anonymous closure> (package:clnapp/model/app_model/list_pays.dart:13:34)
flutter: │ ⛔ #2      MappedListIterable.elementAt (dart:_internal/iterable.dart:415:31)
flutter: │ ⛔ #3      ListIterator.moveNext (dart:_internal/iterable.dart:344:26)
flutter: │ ⛔ #4      new _GrowableList._ofEfficientLengthIterable (dart:core-patch/growable_array.dart:189:27)
flutter: │ ⛔ #5      new _GrowableList.of (dart:core-patch/growable_array.dart:150:28)
flutter: │ ⛔ #6      new List.of (dart:core-patch/array_patch.dart:52:28)
flutter: │ ⛔ #7      ListIterable.toList (dart:_internal/iterable.dart:214:7)
flutter: │ ⛔ #8      new AppListPays.fromJSON (package:clnapp/model/app_model/list_pays.dart:14:12)
flutter: │ ⛔ #9      CLNApi.listPays.<anonymous closure> (package:clnapp/api/cln/cln_client.dart:189:49)
flutter: │ ⛔ #10     RPCClient.call (package:clightning_rpc/src/clightning_rpc.dart:38:22)
flutter: │ ⛔ <asynchronous suspension>
flutter: │ ⛔ #11     _HomeViewState.listPayments (package:clnapp/views/home/home_view.dart:118:22)
flutter: │ ⛔ <asynchronous suspension>
flutter: │ ⛔ #12     _FutureBuilderState._subscribe.<anonymous closure> (package:flutter/src/widgets/async.dart:628:33)
flutter: │ ⛔ <asynchronous suspension>
flutter: │ ⛔ 
flutter: └───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

```

**Comment by Harshit933** [comment]
Exactly which value is it referring to as Null

**Comment by vincenzopalazzo** [comment]
>Exactly which value is it referring to as Null


It is in the stack trace, you need just to debug and consider to take a look the core lightning docs because some field will be null

**Comment by vincenzopalazzo** [comment]
Looks good I tested it, and the only thing that I do not like is the title card!

I would like to add the label of bolt11 (this required to call `decodeinvoice`) what do you think?

Maybe this can be done in a follow-up PR

**Comment by Harshit933** [comment]
Sorry for the replying you late, yes we can have a `decodeInvoice` in the follow up PRs. Should I rebase all the commits for this one or is there any thing that can be further upgraded in this PR?

**Review by vincenzopalazzo — APPROVED** [review]

**Pull request merged**